spark/src/main/scala/com/gu/tableversions/spark/VersionedDataset.scala

```diff
@@ -1,3 +1,131 @@
 package com.gu.tableversions.spark
 
-object VersionedDataset {}
+import java.net.URI
+import java.time.Instant
+
+import cats.effect.IO
+import com.gu.tableversions.core.TableVersions.{TableUpdate, UpdateMessage, UserId}
+import com.gu.tableversions.core._
+import com.gu.tableversions.metastore.Metastore.TableChanges
+import com.gu.tableversions.metastore.{Metastore, VersionPaths}
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+
+/**
+  * Code for writing Spark datasets to storage in a version-aware manner, taking in version information,
+  * using the appropriate paths for storage, and committing version changes.
+  */
+object VersionedDataset {
+
+  implicit class DatasetOps[T](val delegate: Dataset[T])(
+      implicit tableVersions: TableVersions[IO],
+      metastore: Metastore[IO]) {
+
+    /**
+      * Insert the dataset into the given versioned table.
+      *
+      * This emulates the behaviour of Hive inserts in that it will overwrite any partitions present in the dataset,
+      * while leaving other partitions unchanged.
+      *
+      * @return a tuple containing the updated table version information, and a list of the changes that were applied
+      *         to the metastore.
+      */
+    def versionedInsertInto(table: TableDefinition, userId: UserId, message: String)(
+        implicit tableVersions: TableVersions[IO],
+        metastore: Metastore[IO]): (TableVersion, TableChanges) =
+      versionedInsertDatasetIntoTable(delegate, table, userId, message).unsafeRunSync()
+
+  }
+
+  private def versionedInsertDatasetIntoTable[T](
+      dataset: Dataset[T],
+      table: TableDefinition,
+      userId: UserId,
+      message: String)(
+      implicit tableVersions: TableVersions[IO],
+      metastore: Metastore[IO]): IO[(TableVersion, TableChanges)] =
+    for {
+      // Find the partition values in the given dataset
+      datasetPartitions <- IO(VersionedDataset.partitionValues(dataset, table.partitionSchema)(dataset.sparkSession))
+
+      // Get next version numbers for the partitions of the dataset
+      workingVersions <- tableVersions.nextVersions(table.name, datasetPartitions)
+
+      // Resolve the path that each partition should be written to, based on their version
+      partitionPaths = VersionPaths.resolveVersionedPartitionPaths(workingVersions, table.location)
+
+      // Write Spark dataset to the versioned path
+      _ <- IO(VersionedDataset.writeVersionedPartitions(dataset, partitionPaths))
```

Comment:
Do we want to try and do any cleanup if this succeeds but updating the metastore doesn't work? I suppose we probably wouldn't have to since the next time we try and write, the new (hidden) versioned data will be overwritten...?

Reply:
Yes the latter, I'd say. Cleaning up is just more stuff to do that might fail, and I'm not sure it's useful.